qe/drivers.py

```diff
@@ -1707,3 +1732,683 @@ def selection_step(
+        #        self.Hamiltonian_manager = Hamiltonian_manager(
+        #            psi_full, d_one_e_integral, d_two_e_integral, E0, comm
+        #        )
+        # For easier reference, full problem size
+        self.full_size = problem_size
+        # Set local problem size
+        floor = self.full_size // self.world_size
+        ceiling = floor + 1
+        remainder = self.full_size % self.world_size
+        self.distribution = [ceiling] * remainder + [floor] * (self.world_size - remainder)
+        self.local_size = self.distribution[self.rank]  # Size of local chunk of Hamiltonian
+        if self.rank == 0:
+            print(
+                f"Distribution of work: {self.distribution}, Local problem size: {self.local_size}"
+            )
+        # Compute offsets (start of the local section) for all nodes
+        self.offsets = [0] + list(accumulate(self.distribution))
+        del self.offsets[-1]
+        # TODO: Just change one of these. I'm redoing some work in this and dispatch_psi
+        # Throw error if allocation of work is not consistent
+
+    #        assert self.local_size == self.Hamiltonian_manager.local_size
+
+    def parallel_restart(self, dim_S, n_eig, n_newvecs, X_ik, V_ik, W_ik):
+        """Restart Davidson's iteration; resize the trial subspace V_k
+        and its associated data structures. Prevent a significant
+        blow-up of column dimension.
+
+        :param dim_S: column-size of local working variables (V_ik and friends)
+        :param n_eig: number of eigenvalues to look for (minimally allowed dim_S)
+        :param n_newvecs: number of new vectors added at the previous iteration
+        :param X_ik: current Ritz vectors, numpy array
+        :param V_ik, W_ik: local working variables, numpy arrays
+
+        :return new values for dim_S, V_ik, and W_ik following implicit restart
+        """
+        if n_newvecs == 0:
+            V_inew = X_ik[:, :n_eig]
+        else:
+            V_inew = np.c_[X_ik[:, :n_eig], V_ik[:, -n_newvecs:]]
+        # Take leading n_eig Ritz vectors as new guess vectors
+        dim_S = V_inew.shape[1]  # New subspace dimension
+        V_ik = np.zeros((self.local_size, 0), dtype="float")
+        W_ik = np.zeros((self.local_size, 0), dtype="float")
+        # Initialize; normalize first basis vector
+        v_inew = np.array(V_inew[:, 0], dtype="float")
+        v_new = np.zeros(self.full_size, dtype="float")
+        self.comm.Allgatherv(
+            [v_inew, MPI.DOUBLE], [v_new, self.distribution, self.offsets, MPI.DOUBLE]
+        )
+        V_ik = np.c_[V_ik, v_inew / np.linalg.norm(v_new)]
+        for j in range(1, dim_S):
+            # Orthogonalize next vector against previous ones in restart basis
+            v_inew, norm_vnew = self.mgs(V_ik[:, :j], V_inew[:, j])
+            V_ik = np.c_[V_ik, v_inew]  # Update basis
+        n_newvecs = dim_S
+
+        return dim_S, n_newvecs, V_ik, W_ik
```

Comment:
If `n_newvecs` always egual to `dim_S` do we need to return both of them?

Reply:
No, `dim_S` tracks the overall subspace dimension,`n_newvecs` is the no. of trial vectors added at each iteration. So, the function doesn't need to return this. I can simplify it a little and set it outside of the function